File: chapter_3_stacks_and_queues/js/stack.js
/* first in, last out */
class Stack {
  constructor() {
    /* Stack constructor */
    this.stack = []
    this.length = 0
  }

  /* Pushes an element to the top of the stack */
  push(value) {
    this.stack.push(value)
    this.length++
  }

  /* Pops an element off the top of the stack */
  pop() {
    this.length--
    return this.stack.pop()
  }

  /* Determines if a stack is empty */
  isEmpty() {
    return this.length === 0
  }

  /* Peeks at the element at the top of the stack */
  peek() {
    return this.stack[this.length - 1]
  }
}

module.exports = {
  Stack
}

const st = new Stack()
console.log(st) // Stack { stack: [], length: 0 }
st.push(1)
st.push(2)
st.push(4)
console.log(st.peek()) // 4
st.pop()
console.log(st) // Stack { stack: [ 1, 2 ], length: 2 }

File: chapter_3_stacks_and_queues/python/stack.py
class Stack:
    def __init__(self):
        self.stack = []
        self.length = 0

    def push(self, value):
        self.stack.append(value)
        self.length += 1

    def pop(self):
        return self.stack.pop()

    def peek(self):
        return self.stack[self.length - 1]

    def is_empty(self):
        return self.length == 0


if __name__ == "main":
    s = Stack()
    for i in [1, 2, 3, 4, 5]:
        s.push(i)
    print(s.stack)  # [1, 2, 3, 4, 5]
    print(s.pop())  # 5
    print(s.peek())  # 4
    print(s.stack)  # [1, 2, 3, 4]
